### PR TITLE
WEA-5 Implementation of Retrofit library to fetch weather info

### PR DESCRIPTION
**Pull request by ayush-shta**
Also implemented MVP structure to some extent

### CODE REVIEW

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java [hunk 1]
```diff
@@ -9,16 +9,22 @@
...
 
-public class MainActivity extends BaseActivity {
+public class MainActivity extends BaseActivity implements MainContract.View {
 
   @BindView(R.id.layout_drawer) DrawerLayout drawerLayout;
   @BindView(R.id.nav_view) NavigationView navigationView;
   @BindView(R.id.toolbar) Toolbar toolbar;
+  @BindView(R.id.txt_main) TextView textView;
+
+  ActionBar actionbar;
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: Could this be done in base activity and the derived activity only provided the title for the action bar.

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java [hunk 2]
```diff
@@ -32,6 +38,11 @@
     setActionBar();
 
     setNavigationView();
+
+    presenter = new MainPresenter(this);
+    presenter.setView(this);
+
+    presenter.apiCall();
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: presenter.fetchWeatherDetail would be nice

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java [hunk 3]
```diff
@@ -52,13 +63,11 @@
   private void setActionBar() {
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: this code should be in base activity

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java [hunk 4]
```diff
@@ -77,4 +86,8 @@ private void setNavigationView() {
           }
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: can we use lambda expressions here?

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java [hunk 5]
```diff
@@ -77,4 +86,8 @@ private void setNavigationView() {
...
         });
   }
+
+  @Override public void setText(String s) {
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: setText() is very generic. The methods should be particular like setTemperature or set/showError.

File: app/src/main/java/com/ayush/weatherapp/main/MainContract.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  }
+
+  interface Presenter extends BasePresenter {
+    void showProgressDialog();
+
+    void hideProgressDialog();
+
+    void setView(View view);
+
+    void apiCall();
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainContract.java]: what Api to call?

File: app/src/main/java/com/ayush/weatherapp/main/MainContract.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.ayush.weatherapp.main;
+
+import com.ayush.weatherapp.BasePresenter;
+
+public interface MainContract {
+  interface View {
+    void setText(String s);
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainContract.java]: what text to set?

File: app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import android.support.annotation.NonNull;
+import com.ayush.weatherapp.main.pojo.Currently;
+import com.ayush.weatherapp.main.pojo.Forecast;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class MainPresenter implements MainContract.Presenter {
+
+  private static ProgressDialog progressDialog;
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java]: Presenter should not contain any view(android.View) objects. This should be in the activity. Context cannot be helped.

File: app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+        Currently currently = forecast.getCurrently();
+
+        view.setText(currently.getSummary());
+
+        hideProgressDialog();
+      }
+
+      @Override public void onFailure(@NonNull Call<Forecast> call, @NonNull Throwable t) {
+        call.cancel();
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java]: do we need to cancel on failure?

File: app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    String requestString = TEST_LATITUDE + "," + TEST_LONGITUDE;
+
+    Call<Forecast> forecastCall = apiInterface.getForecast(requestString);
+
+    forecastCall.enqueue(new Callback<Forecast>() {
+      @Override
+      public void onResponse(@NonNull Call<Forecast> call, @NonNull Response<Forecast> response) {
+        Forecast forecast = response.body();
+
+        Currently currently = forecast.getCurrently();
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java]: Class name Currently is not appropriate. The name should indicate what it contains or does as far as possible

File: app/src/main/java/com/ayush/weatherapp/main/pojo/Daily.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
+package com.ayush.weatherapp.main.pojo;
+
+import com.google.gson.annotations.SerializedName;
+import java.util.List;
+
+public class Daily {
+  @SerializedName("summary") private String summary;
+  @SerializedName("icon") private String icon;
+  @SerializedName("data") private List<DailyData> dailyDataList = null;
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/main/pojo/Daily.java]: = null not required its explicitly null

File: app/src/main/res/layout/main.xml [hunk 12]
```diff
@@ -5,8 +5,11 @@
...
 
   <TextView
+      android:id="@+id/txt_main"
       android:layout_width="wrap_content"
       android:layout_height="wrap_content"
+      android:layout_centerInParent="true"
+      android:inputType="textMultiLine"
       android:text="main"
       />
 </RelativeLayout>
```
Comment (subash-poudel) [inline app/src/main/res/layout/main.xml]: new line

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -7,4 +7,5 @@
 
   <string name="celsius">°C</string>
   <string name="fahrenheit">°F</string>
```
Comment (subash-poudel) [inline app/src/main/res/values/strings.xml]: Have a look at xliff:g
https://github.com/marcelb/android-simple-weather/blob/master/app/src/main/res/values/strings.xml

File: app/src/main/res/layout/main.xml (hunk 12 above)
Comment (ayush-shta) [reply app/src/main/res/layout/main.xml]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/pojo/Daily.java (hunk 11 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/pojo/Daily.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java (hunk 10 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainContract.java (hunk 7 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainContract.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainContract.java (hunk 6 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainContract.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java (hunk 5 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java (hunk 4 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java (hunk 2 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java (hunk 8 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java (hunk 3 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainActivity.java (hunk 1 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: :ok:

Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java (hunk 9 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/main/MainPresenter.java]: :ok:

**Pull request merged**